hardware_interface/test/test_resource_manager_prepare_perform_switch.cpp

```diff
@@ -0,0 +1,389 @@
+    ASSERT_TRUE(rm_->perform_command_mode_switch(empty_keys, empty_keys));
+
+    // Set both HW to ACTIVE to claim interfaces. There should stay persistent because we are not
+    // cleaning them for now, so this is a good way to keep the access and "f* the system"
+    set_components_state(
+      *rm_, {"TestSystemCommandModes"}, lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, "active");
+    set_components_state(
+      *rm_, {"TestActuatorHardware"}, lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, "active");
+    // State to get feedback how many times "prepare_for_switch" is called
+    claimed_system_acceleration_state_ = std::make_unique<hardware_interface::LoanedStateInterface>(
+      rm_->claim_state_interface("joint1/acceleration"));
+    claimed_actuator_position_state_ = std::make_unique<hardware_interface::LoanedStateInterface>(
+      rm_->claim_state_interface("joint3/position"));
+  }
+
+  void preconfigure_components(
+    const uint8_t system_state_id, const std::string syste_state_name,
+    const uint8_t actuator_state_id, const std::string actuator_state_name)
+  {
+    set_components_state(*rm_, {"TestSystemCommandModes"}, system_state_id, syste_state_name);
+    set_components_state(*rm_, {"TestActuatorHardware"}, actuator_state_id, actuator_state_name);
+
+    auto status_map = rm_->get_components_status();
+    EXPECT_EQ(status_map["TestSystemCommandModes"].state.id(), system_state_id);
+    EXPECT_EQ(status_map["TestActuatorHardware"].state.id(), actuator_state_id);
+  }
+
+  std::unique_ptr<TestableResourceManager> rm_;
+
+  std::unique_ptr<hardware_interface::LoanedStateInterface> claimed_system_acceleration_state_;
+  std::unique_ptr<hardware_interface::LoanedStateInterface> claimed_actuator_position_state_;
+
+  // Scenarios defined by example criteria
+  std::vector<std::string> empty_keys = {};
+  std::vector<std::string> non_existing_keys = {"elbow_joint/position", "should_joint/position"};
+  std::vector<std::string> legal_keys_system = {"joint1/position", "joint2/position"};
+  std::vector<std::string> legal_keys_actuator = {"joint3/position"};
+};
+
+// System  : ACTIVE
+// Actuator: UNCONFIGURED
+TEST_F(
+  ResourceManagerPreparePerformTest,
+  when_system_active_and_actuator_unconfigured_expect_system_passing)
+{
+  preconfigure_components(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, "active",
+    lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED, "unconfigured");
+
+  // Default behavior for empty key lists, e.g., when a Broadcaster is activated
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(empty_keys, empty_keys));
+  // empty call can be done at any time and it doesn't propagates to HW components - always true
+  EXPECT_TRUE(rm_->perform_command_mode_switch(empty_keys, empty_keys));
+
+  // When TestSystemCommandModes is ACTIVE expect OK
+  EXPECT_TRUE(rm_->prepare_command_mode_switch(legal_keys_system, legal_keys_system));
```

Comment:
This is just for the test purposes. Nevertheless, I think this might happen when you replace one controller with another that uses the same interfaces.